# Atividade 1 — Exercícios com os comandos básicos de pandas para EDA


Responda as perguntas abaixo no notebook (cada resposta deve ter ao menos uma célula de código que produza o output e, logo abaixo, uma célula de texto com a sua interpretação).

## 1) Carregue o dataset e mostre as primeiras 10 linhas.

In [4]:
import pandas as pd
url = "https://drive.google.com/uc?export=download&id=1eJ-L0MnJxHi929fZrVOKxH7Mrw5eTGpk"
Df = pd.read_csv(url)
Df.head(10)


/tmp/ipykernel_539/1694070970.py:3: DtypeWarning: Columns (0,3,5) have mixed types. Specify dtype option on import or set low_memory=False.
  Df = pd.read_csv(url)


,Código da Emenda,Ano da Emenda,Tipo de Emenda,Código do Autor da Emenda,Nome do Autor da Emenda,Número da emenda,Localidade de aplicação do recurso,Código Município IBGE,Município,Código UF IBGE,...,Código Ação,Nome Ação,Código Plano Orçamentário,Nome Plano Orçamentário,Valor Empenhado,Valor Liquidado,Valor Pago,Valor Restos A Pagar Inscritos,Valor Restos A Pagar Cancelados,Valor Restos A Pagar Pagos
0,Sem informação,2014,Emenda Individual - Transferências com Finalid...,S/I,Sem informação,S/I,ALVORADA - RS,4300604,ALVORADA,4300000,...,1D73,APOIO A POLITICA NACIONAL DE DESENVOLVIMENTO U...,0000,APOIO A POLITICA NACIONAL DE DESENVOLVIMENTO U...,"700000,00","0,00","0,00","0,00","690900,00","9100,00"
1,Sem informação,2014,Emenda Individual - Transferências com Finalid...,S/I,Sem informação,S/I,SANTA ISABEL DO PARÁ - PA,1506500,SANTA ISABEL DO PARÁ,1500000,...,10V0,APOIO A PROJETOS DE INFRAESTRUTURA TURISTICA,0000,APOIO A PROJETOS DE INFRAESTRUTURA TURISTICA -...,"487500,00","0,00","0,00","0,00","487500,00","0,00"
2,Sem informação,2014,Emenda Individual - Transferências com Finalid...,S/I,Sem informação,S/I,BARRA BONITA - SC,4202099,BARRA BONITA,4200000,...,8581,ESTRUTURACAO DA REDE DE SERVICOS DE ATENCAO PR...,0000,ESTRUTURACAO DA REDE DE SERVICOS DE ATENCAO BA...,"100000,00","0,00","0,00","0,00","0,00","100000,00"
3,Sem informação,2014,Emenda Individual - Transferências com Finalid...,S/I,Sem informação,S/I,ARARAQUARA - SP,3503208,ARARAQUARA,3500000,...,10V0,APOIO A PROJETOS DE INFRAESTRUTURA TURISTICA,0000,APOIO A PROJETOS DE INFRAESTRUTURA TURISTICA -...,"243750,00","0,00","0,00","0,00","243750,00","0,00"
4,Sem informação,2014,Emenda Individual - Transferências com Finalid...,S/I,Sem informação,S/I,CAMPO LARGO - PR,4104204,CAMPO LARGO,4100000,...,8581,ESTRUTURACAO DA REDE DE SERVICOS DE ATENCAO PR...,0000,ESTRUTURACAO DA REDE DE SERVICOS DE ATENCAO BA...,"1658333,94","0,00","0,00","0,00","527200,00","1131133,94"
5,Sem informação,2014,Emenda Individual - Transferências com Finalid...,S/I,Sem informação,S/I,CAMPINAS DO SUL - RS,4303806,CAMPINAS DO SUL,4300000,...,10V0,APOIO A PROJETOS DE INFRAESTRUTURA TURISTICA,0000,APOIO A PROJETOS DE INFRAESTRUTURA TURISTICA -...,"243750,00","0,00","0,00","0,00","0,00","243750,00"
6,Sem informação,2014,Emenda Individual - Transferências com Finalid...,S/I,Sem informação,S/I,MORENO - PE,2609402,MORENO,2600000,...,20RP,APOIO A INFRAESTRUTURA PARA A EDUCACAO BASICA,0000,INFRAESTRUTURA PARA A EDUCACAO BASICA - DESPES...,"700000,00","0,00","0,00","0,00","700000,00","0,00"
7,Sem informação,2014,Emenda Individual - Transferências com Finalid...,S/I,Sem informação,S/I,NOVA FLORESTA - PB,2510105,NOVA FLORESTA,2500000,...,8581,ESTRUTURACAO DA REDE DE SERVICOS DE ATENCAO PR...,0000,ESTRUTURACAO DA REDE DE SERVICOS DE ATENCAO BA...,"278700,00","0,00","0,00","0,00","0,00","278700,00"
8,Sem informação,2014,Emenda Individual - Transferências com Finalid...,S/I,Sem informação,S/I,SANTIAGO DO SUL - SC,4215695,SANTIAGO DO SUL,4200000,...,20ZV,FOMENTO AO SETOR AGROPECUARIO,0000,FOMENTO AO SETOR AGROPECUARIO - DESPESAS DIVERSAS,"300000,00","300000,00","300000,00","0,00","0,00","0,00"
9,Sem informação,2014,Emenda Individual - Transferências com Finalid...,S/I,Sem informação,S/I,CIANORTE - PR,4105508,CIANORTE,4100000,...,1D73,APOIO A POLITICA NACIONAL DE DESENVOLVIMENTO U...,0000,APOIO A POLITICA NACIONAL DE DESENVOLVIMENTO U...,"245850,00","0,00","0,00","0,00","0,00","245850,00"


Explique em uma frase o que você observa nas primeiras 10 linhas.
As primeiras 10 linhas mostram registros de emendas individuais do ano de 2014, adicionando informações como a localização pra onde os recurso foram direcionados, a finalidade de tais orçamentos e como foi feita a execução desse ponto de vista financeiro.


## 2) Quantas linhas e colunas o dataset possui?

In [5]:
Df.shape

(88596, 28)

Os números representam a quantidade de linhas (registro) e colunas (característica/variável) do dataset. São 88596 linhas e 28 colunas.

## 3) Liste os nomes das colunas e descreva brevemente o que você entende que cada coluna representa.

In [ ]:
Df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 88596 entries, 0 to 88595
Data columns (total 28 columns):
 #   Column                              Non-Null Count  Dtype 
---  ------                              --------------  ----- 
 0   Código da Emenda                    88596 non-null  object
 1   Ano da Emenda                       88596 non-null  int64 
 2   Tipo de Emenda                      88596 non-null  object
 3   Código do Autor da Emenda           88596 non-null  object
 4   Nome do Autor da Emenda             88596 non-null  object
 5   Número da emenda                    88596 non-null  object
 6   Localidade de aplicação do recurso  88596 non-null  object
 7   Código Município IBGE               88596 non-null  object
 8   Município                           88596 non-null  object
 9   Código UF IBGE                      88596 non-null  int64 
 10  UF                                  88596 non-null  object
 11  Região                              88596 non-null  ob


Com o comando df.info pode visualizar que este conjunto de dados contém 88.596 registros e 28 colunas estruturado da seguinte maneira:
- Identificação e autoria 
- Geolocalização
- Classificação Orrçamentária
- Ciclo Financeiro.

## 4) Existem valores nulos? quantos?


In [12]:
nulos = Df.isnull().sum()
nulos

,0
Código da Emenda,0
Ano da Emenda,0
Tipo de Emenda,0
Código do Autor da Emenda,0
Nome do Autor da Emenda,0
Número da emenda,0
Localidade de aplicação do recurso,0
Código Município IBGE,0
Município,0
Código UF IBGE,0


O dataset não apresenta dados ausentes (nulos).

## 5) Mostre os tipos das variáveis (se int, string, boolean, float...) e analise se alguma variável parece ter o tipo errado. Faça isso da maneira mais objetiva e otimizada possível.

In [13]:
Df.dtypes

,0
Código da Emenda,object
Ano da Emenda,int64
Tipo de Emenda,object
Código do Autor da Emenda,object
Nome do Autor da Emenda,object
Número da emenda,object
Localidade de aplicação do recurso,object
Código Município IBGE,object
Município,object
Código UF IBGE,int64


Se alguma coluna estiver com o tipo errado, escreva abaixo qual é e por que isso pode atrapalhar análises.

Divergência de Tipos: As colunas de execução financeira (ex: Valor Empenhado, Valor Pago) estão classificadas como object (string), quando deveriam ser float64 (numéricas)

Impacto Técnico: o Phyton entende que tais valores são texto, esta configuração inviabiliza cálculos matemáticos, agregações estatísticas e a plotagem correta de gráficos.

## 6) Usando a coluna `Ano da Emenda` como base, descubra até que ano o dataset está atualizado (mostre as 8 primeiras linhas do df ordenado desc).

In [19]:
Df = Df.sort_values(by='Ano da Emenda', ascending=False)
Df.head(8)


,Código da Emenda,Ano da Emenda,Tipo de Emenda,Código do Autor da Emenda,Nome do Autor da Emenda,Número da emenda,Localidade de aplicação do recurso,Código Município IBGE,Município,Código UF IBGE,...,Código Ação,Nome Ação,Código Plano Orçamentário,Nome Plano Orçamentário,Valor Empenhado,Valor Liquidado,Valor Pago,Valor Restos A Pagar Inscritos,Valor Restos A Pagar Cancelados,Valor Restos A Pagar Pagos
88564,202541820003,2025,Emenda Individual - Transferências Especiais,4182,ZEQUINHA MARINHO,3,MÚLTIPLO,Sem informação,Múltiplo,1500000,...,0EC2,TRANSFERENCIAS ESPECIAIS,0000,TRANSFERENCIAS ESPECIAIS,"12161000,00","12039390,00","12039390,00","0,00","0,00","0,00"
88578,202544940001,2025,Emenda Individual - Transferências com Finalid...,4494,ZEZINHO BARBARY,1,Nacional,Sem informação,Sem informação,-1,...,2726,PREVENCAO E REPRESSAO AO TRAFICO ILICITO DE DR...,0000,PREVENCAO E REPRESSAO AO TRAFICO ILICITO DE DR...,"478233,30","0,00","0,00","0,00","0,00","0,00"
88595,202544840009,2025,Emenda Individual - Transferências com Finalid...,4484,ZUCCO,9,MÚLTIPLO,Sem informação,Múltiplo,4300000,...,20JP,DESENVOLVIMENTO DE ATIVIDADES E APOIO A PROGRA...,0000,DESENVOLVIMENTO DE ATIVIDADES E APOIO A PROGRA...,"1800000,00","36000,00","0,00","0,00","0,00","0,00"
88594,202544840008,2025,Emenda Individual - Transferências com Finalid...,4484,ZUCCO,8,Nacional,Sem informação,Sem informação,-1,...,2723,"POLICIAMENTO, FISCALIZACAO, ENFRENTAMENTO A CR...",0000,"POLICIAMENTO, FISCALIZACAO, ENFRENTAMENTO A CR...","497151,34","0,00","0,00","0,00","0,00","0,00"
88577,202541820017,2025,Emenda Individual - Transferências com Finalid...,4182,ZEQUINHA MARINHO,17,REDENÇÃO - PA,1506138,REDENÇÃO,1500000,...,219G,ESTRUTURACAO DA REDE DE SERVICOS E FORTALECIME...,0000,ESTRUTURACAO DA REDE DE SERVICOS E FORTALECIME...,"370000,00","370000,00","370000,00","0,00","0,00","0,00"
88576,202541820016,2025,Emenda Individual - Transferências com Finalid...,4182,ZEQUINHA MARINHO,16,PARÁ (UF),Sem informação,Sem informação,1500000,...,8535,ESTRUTURACAO DE UNIDADES DE ATENCAO ESPECIALIZ...,0000,ESTRUTURACAO DE UNIDADES DE ATENCAO ESPECIALIZ...,"600000,00","0,00","0,00","0,00","0,00","0,00"
88575,202541820015,2025,Emenda Individual - Transferências com Finalid...,4182,ZEQUINHA MARINHO,15,PARÁ (UF),Sem informação,Sem informação,1500000,...,20Z1,QUALIFICACAO SOCIAL E PROFISSIONAL DE TRABALHA...,0000,QUALIFICACAO SOCIAL E PROFISSIONAL DE TRABALHA...,"500000,00","0,00","0,00","0,00","0,00","0,00"
88574,202541820014,2025,Emenda Individual - Transferências com Finalid...,4182,ZEQUINHA MARINHO,14,MÚLTIPLO,Sem informação,Múltiplo,1500000,...,00SL,APOIO A IMPLANTACAO E MODERNIZACAO DE INFRAEST...,0000,APOIO A IMPLANTACAO E MODERNIZACAO DE INFRAEST...,"500000,00","0,00","0,00","0,00","0,00","0,00"


Explique: qual o ano mais recente presente no dataset?

O dataset está atualizado até o ano de 2025.


## 7) execute um  `describe()` e explique uma conclusão simples que você consegue tirar a partir das estatísticas (se algumas colunas deveriam aparecer no describe mas não aparecem, indique o motivo que você acredita que esteja causando esse problema).

In [21]:
Df.describe()

,Ano da Emenda,Código UF IBGE
count,88596.000000,8.859600e+04
mean,2019.155109,2.943594e+06
std,3.409323,1.264884e+06
min,2014.000000,-1.400000e+01
25%,2016.000000,2.300000e+06
50%,2019.000000,3.100000e+06
75%,2022.000000,3.500000e+06
max,2025.000000,5.300000e+06


O describe() mostra que o dataset cobrem um período de 11 anos (de 2014 a 2025). A média e a mediana (50%) estão em torno de 2019, o que indica que o volume de emendas possui uma boa distribuição temporal (2014-2025), mas revela uma falha na tipagem: as variáveis financeiras não aparecem no resumo estatístico por estarem classificadas como object (texto).

## 8) Mostre as últimas 5 linhas do dataset e escreva se a última linha parece completa (sem muitos nulos) ou não.

In [22]:
Df.tail()

,Código da Emenda,Ano da Emenda,Tipo de Emenda,Código do Autor da Emenda,Nome do Autor da Emenda,Número da emenda,Localidade de aplicação do recurso,Código Município IBGE,Município,Código UF IBGE,...,Código Ação,Nome Ação,Código Plano Orçamentário,Nome Plano Orçamentário,Valor Empenhado,Valor Liquidado,Valor Pago,Valor Restos A Pagar Inscritos,Valor Restos A Pagar Cancelados,Valor Restos A Pagar Pagos
7466,Sem informação,2014,Emenda Individual - Transferências com Finalid...,S/I,Sem informação,S/I,BATALHA - PI,2201507,BATALHA,2200000,...,7K66,APOIO A PROJETOS DE DESENVOLVIMENTO SUSTENTAVE...,0000,APOIO A PROJETOS DE DESENVOLVIMENTO SUSTENTAVE...,"101650,00","101650,00","0,00","0,00","0,00","101650,00"
7467,Sem informação,2014,Emenda Individual - Transferências com Finalid...,S/I,Sem informação,S/I,QUARAÍ - RS,4315305,QUARAÍ,4300000,...,8581,ESTRUTURACAO DA REDE DE SERVICOS DE ATENCAO PR...,0000,ESTRUTURACAO DA REDE DE SERVICOS DE ATENCAO BA...,"150000,00","0,00","0,00","0,00","0,00","150000,00"
7468,Sem informação,2014,Emenda Individual - Transferências com Finalid...,S/I,Sem informação,S/I,MARIZÓPOLIS - PB,2509156,MARIZÓPOLIS,2500000,...,5450,IMPLANTACAO E MODERNIZACAO DE INFRAESTRUTURA P...,0000,IMPLANTACAO E MODERNIZACAO DE INFRAESTRUTURA P...,"390000,00","0,00","0,00","0,00","0,00","390000,00"
7469,Sem informação,2014,Emenda Individual - Transferências com Finalid...,S/I,Sem informação,S/I,ALVORADA D'OESTE - RO,1100346,ALVORADA D'OESTE,1100000,...,4525,APOIO A MANUTENCAO DE UNIDADES DE SAUDE,0000,APOIO A MANUTENCAO DE UNIDADES DE SAUDE - DESP...,"100000,00","0,00","0,00","0,00","0,00","100000,00"
7459,Sem informação,2014,Emenda Individual - Transferências com Finalid...,S/I,Sem informação,S/I,DOM FELICIANO - RS,4306502,DOM FELICIANO,4300000,...,1D73,APOIO A POLITICA NACIONAL DE DESENVOLVIMENTO U...,0000,APOIO A POLITICA NACIONAL DE DESENVOLVIMENTO U...,"245850,00","0,00","0,00","0,00","0,00","245850,00"


Interprete a completude da última linha.

Embora não existam buracos, a última linha, por exemplo, contém termos como "Sem informação" e "S/I". Para o Python, isso conta como "preenchido" (porque tem texto escrito), mas para análise, isso é um dado faltante mascarado.

## 9) existem valores duplicados no dataframe? Caso positivo, quantos?

In [23]:
Duplicatas = Df.duplicated().sum()
print(Duplicatas)

0


Não existem valores duplicados no dataframe.